track_analyzing/track_analyzer/cmd_table.cpp

```diff
@@ -19,88 +20,198 @@
   }
 
-  uint32_t GetType(FeatureType & feature) const
+  struct Type
+  {
+    bool operator<(Type const & rhs) const
+    {
+      if (m_hwType != rhs.m_hwType)
+        return m_hwType < rhs.m_hwType;
+
+      return m_surfaceType < rhs.m_surfaceType;
+    }
+
+    bool operator==(Type const & rhs) const
+    {
+      return m_hwType == rhs.m_hwType && m_surfaceType == rhs.m_surfaceType;
+    }
+
+    bool operator!=(Type const & rhs) const
+    {
+      return !(*this == rhs);
+    }
+
+    uint32_t m_hwType = 0;
+    uint32_t m_surfaceType = 0;
+  };
+
+  Type GetType(FeatureType & feature) const
   {
+    Type ret;
     feature::TypesHolder holder(feature);
-    for (uint32_t type : m_tags)
+    for (uint32_t type : m_hwtags)
     {
       if (holder.Has(type))
-        return type;
+      {
+        ret.m_hwType = type;
+        break;
+      }
     }
 
-    return 0;
+    for (uint32_t type : m_surfaceTags)
+    {
+      if (holder.Has(type))
+      {
+        ret.m_surfaceType = type;
+        break;
+      }
+    }
+
+    return ret;
   }
 
 private:
-  vector<uint32_t> m_tags;
+  vector<uint32_t> m_hwtags;
+  vector<uint32_t> m_surfaceTags;
+};
+
+struct RoadInfo
+{
+  bool operator==(RoadInfo const & rhs) const
+  {
+    return m_type == rhs.m_type && m_isCityRoad == rhs.m_isCityRoad && m_isOneWay == rhs.m_isOneWay;
+  }
+
+  bool operator!=(RoadInfo const & rhs) const
+  {
+    return !(*this == rhs);
+  }
+
+  bool operator<(RoadInfo const & rhs) const
+  {
+    if (m_type != rhs.m_type)
+      return m_type < rhs.m_type;
+
+    if (m_isCityRoad != rhs.m_isCityRoad)
+      return !m_isCityRoad;
+
+    return !m_isOneWay;
+  }
+
+  CarModelTypes::Type m_type;
+  bool m_isCityRoad = false;
+  bool m_isOneWay = false;
 };
 
 class MoveType final
 {
 public:
   MoveType() = default;
 
-  MoveType(uint32_t roadType, traffic::SpeedGroup speedGroup)
-    : m_roadType(roadType), m_speedGroup(speedGroup)
+  MoveType(RoadInfo const & roadType, traffic::SpeedGroup speedGroup, bool isDayTime)
+    : m_roadInfo(roadType), m_speedGroup(speedGroup), m_isDayTime(isDayTime)
   {
   }
 
   bool operator==(MoveType const & rhs) const
   {
-    return m_roadType == rhs.m_roadType && m_speedGroup == rhs.m_speedGroup;
+    return m_roadInfo == rhs.m_roadInfo && m_speedGroup == rhs.m_speedGroup && m_isDayTime == rhs.m_isDayTime;
   }
 
   bool operator<(MoveType const & rhs) const
   {
-    if (m_roadType != rhs.m_roadType)
-      return m_roadType < rhs.m_roadType;
+    if (m_roadInfo != rhs.m_roadInfo)
+      return m_roadInfo < rhs.m_roadInfo;
+
+    if (m_speedGroup != rhs.m_speedGroup)
+      return m_speedGroup < rhs.m_speedGroup;
+
+    return !m_isDayTime;
+  }
 
-    return m_speedGroup < rhs.m_speedGroup;
+  bool IsValid() const
+  {
+    return m_roadInfo.m_type.m_hwType != 0 &&
+           m_roadInfo.m_type.m_surfaceType != 0 &&
+           m_speedGroup == traffic::SpeedGroup::G5;
```

Comment:
Please add a comment why only G5 is used here.

Reply:
ok